### ISSUE TITLE
Span finalizer can cause a memory leak with G1

### ISSUE THREAD

**Issue by JonasKunz** [open-telemetry/opentelemetry-java#1074]
Hi,

We observed an issue with the `finalize` method of the span implementation of the SDK. When under high CPU load, the finalizer can prevent the garbage collection of spans, leading to a highly increased heap consumption.

The root cause for the problem is the following finalizer:
https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java#L522
Finalizers block objects from being GC'd until they are executed. It seems like under high CPU load the JDK 11 G1 Collector with default settings decides to postpone the execution of the span's finalizer, causing the span to block heap memory.

We observed and analyzed the problem when using OpenCensus, which has exactly the same finalizer for spans, therefore we are reporting this issue here. Our JavaAgent internally uses OC to collect trace data, which is where the issue occured. After some heap dump analysis we did an experiment:
We removed the finalizer from the span implementation and changed nothing else. This caused the issue to disappear. For the full details of the problem and the analysis please have a look at the corresponding [issue](https://github.com/inspectIT/inspectit-ocelot/issues/564) at our repo.

The quickfix for this problem would be to simply remove the finalizer, which however also removes the log messages for non-ended spans.
The "correct" fix would be to use `WeakReferences` alongside with a [ReferenceQueue](https://docs.oracle.com/javase/7/docs/api/java/lang/ref/ReferenceQueue.html) to detect when non-ended spans are GC'd, as `WeakReference`s do not block the object they are referring to from being freed.

**Comment by Oberon00** [comment]
We could implement this as providing an optional WarnOnLeakedSpan SpanProcessor that would maintain that WeakReference queue and print the warnings. Then the finalizer could be removed.

**Comment by jkwatson** [comment]
Decision from the java SIG meeting today: We're going to remove the finalizer ASAP. And, I'll create an issue to track creating a span processor that will do what @Oberon00 suggests that can be used for checking for non-finished spans.

**Comment by bogdandrutu** [comment]
@jkwatson let's fix the issue first then do the SpanProcessor because this may affect users.

**Comment by carlosalberto** [comment]
@bogdandrutu To be clear: implement a solution like the one suggested by @Oberon00 and only after that, remove the finalizer?

**Comment by Oberon00** [comment]
@bogdandrutu I read it the other way round: Removing the finalizer has priority and the Processor can be implemented as soon as someone has time.

**Comment by jkwatson** [comment]
Issue created to create the leak-warning span processor:  #1084

**Comment by jkwatson** [comment]
@JonasKunz we've got the finalizer removed. Are you ok with us closing the issue?

**Comment by JonasKunz** [comment]
Sure! I really like the quick actions taken here :)
